pkg/webhook/generator_test.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/spf13/afero"
+)
+
+var expected = map[string]string{
+	"config/webhook/webhookmanifests.yaml": `apiVersion: admissionregistration.k8s.io/v1beta1
+kind: MutatingWebhookConfiguration
+metadata:
+  annotations:
+    admissionwebhook.alpha.kubebuilder.io/ca-secret-name: webhook-cert
+  creationTimestamp: null
+  name: test-mutating-webhook-cfg
+webhooks:
+- clientConfig:
+    service:
+      name: webhook-service
+      namespace: test-system
+      path: /bar
+  failurePolicy: Fail
+  name: bar-webhook
+  namespaceSelector:
+    matchExpressions:
+    - key: control-plane
+      operator: DoesNotExist
+  rules:
+  - apiGroups:
+    - apps
+    operations:
+    - CREATE
+    - UPDATE
+    resources:
+    - deployments
+---
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  annotations:
+    admission.alpha.kubebuilder.io/ca-secret-name: webhook-cert
+  creationTimestamp: null
+  name: test-validating-webhook-cfg
+webhooks:
+- clientConfig:
+    service:
+      name: webhook-service
+      namespace: test-system
+      path: /baz
+  failurePolicy: Ignore
+  name: baz-webhook
+  namespaceSelector:
+    matchExpressions:
+    - key: control-plane
+      operator: DoesNotExist
+  rules:
+  - apiGroups:
+    - crew
+    apiVersions:
+    - v1
+    operations:
+    - DELETE
+    resources:
+    - firstmates
+---
+apiVersion: v1
+kind: Service
+metadata:
+  annotations:
+    service.alpha.kubebuilder.io/serving-cert-secret-name: webhook-cert
+  creationTimestamp: null
+  name: webhook-service
+  namespace: test-system
+spec:
+  ports:
+  - port: 443
+    targetPort: 7890
+  selector:
+    app: webhook-server
+status:
+  loadBalancer: {}
+`,
+	"config/default/manager_patch.yaml": `apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: controller-manager
+spec:
+  template:
+    metadata:
+      labels:
+        app: webhook-server
+      containers:
+      - name: webhook-server-container
+        ports:
+        - containerPort: 7890
+          name: webhook-server
+          protocol: TCP
+        volumeMounts:
+        - mountPath: /tmp/test-cert
+          name: cert
+          readOnly: true
+      volumes:
+      - name: cert
+        secret:
+          defaultMode: 420
+          secretName: webhook-secret
```

Comment:
Two concerns:
 - Looks like input for CA injections (secret-name, service name etc.) is split across the objects. It might make the cert controller implementation difficult. I am wondering if we can combine all these and apply it only on field. Annotation value could be then in JSON ? 
 - webhook-secret is hardcoded. We should think of a way to override it (by default ?)

Reply:
> Looks like input for CA injections (secret-name, service name etc.) is split across the objects. It might make the cert controller implementation difficult.

For generating cert, everything it needs is in the service object.
For syncing CA, everything it needs is in the WebhookConfiguration object.
It should not make the controller hard to implement. Instead, it should be quite easy to implement it :)

> webhook-secret is hardcoded.

This one is not.

Reply:
Got it. You are right, those are decoupled annotations, so it should be simpler :)